fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/runners/TransactionalRunnerTest.java

```diff
@@ -0,0 +1,414 @@
+                return context.getReadVersionAsync();
+            };
+            try (TransactionalRunner runner = defaultTransactionalRunner()) {
+                firstReadVersion = runner.runAsync(false, getReadVersionWithWriteConflict).join();
+            }
+
+            FDBDatabase.WeakReadSemantics weakReadSemantics = new FDBDatabase.WeakReadSemantics(
+                    firstReadVersion, Long.MAX_VALUE, true);
+            final FDBRecordContextConfig.Builder contextConfigBuilder = FDBRecordContextConfig.newBuilder()
+                    .setWeakReadSemantics(weakReadSemantics);
+            try (TransactionalRunner runner = new TransactionalRunner(database, contextConfigBuilder)) {
+                assertEquals(firstReadVersion, runner.runAsync(false, getReadVersionWithWriteConflict).join());
+                final Long newReadVersion = runner.runAsync(true, getReadVersionWithWriteConflict).join();
+                assertNotEquals(firstReadVersion, newReadVersion);
+                assertEquals(newReadVersion, runner.runAsync(false, getReadVersionWithWriteConflict).join());
+
+                assertConflicts(runner.runAsync(false, context -> {
+                    context.ensureActive().addReadConflictKey(key); // will cause conflict
+                    context.ensureActive().addWriteConflictKey(key);
+                    return CompletableFuture.completedFuture("ignored");
+                }));
+
+                assertEquals("boxes", runner.runAsync(true, context -> {
+                    context.ensureActive().addReadConflictKey(key); // no conflict because we cleared the weak read semantics
+                    context.ensureActive().addWriteConflictKey(key);
+                    return CompletableFuture.completedFuture("boxes");
+                }).join());
+
+                assertConflicts(runner.runAsync(false, context -> {
+                    context.ensureActive().addReadConflictKey(key); // conflicts because we are reusing the weak read semantics
+                    context.ensureActive().addWriteConflictKey(key);
+                    return CompletableFuture.completedFuture("ignored");
+                }));
+            }
+        } finally {
+            database.setTrackLastSeenVersionOnRead(tracksReadVersions);
+            database.setTrackLastSeenVersionOnCommit(tracksCommitVersions);
+        }
+    }
+
+    @Test
+    void closesContexts() {
+        List<FDBRecordContext> contexts = new ArrayList<>();
+        AtomicInteger completed = new AtomicInteger();
+        try (TransactionalRunner runner = defaultTransactionalRunner()) {
+            for (int i = 0; i < 20; i++) {
+                runner.runAsync(false, context -> {
+                    contexts.add(context);
+                    return MoreAsyncUtil.delayedFuture(60, TimeUnit.SECONDS)
+                            // delayed future isn't guaranteed, so we need to make sure at least some delayed past the close
```

Comment:
Should this test just return a `new CompletableFuture()` (i.e., a future that never completes)?